File: ruby/trab3/bissecao.rb
# encoding: utf-8
# Algoritmo BISSEÇÃO
require "./base"

Calcnum::algoritmo :bissecao do |xi, xf|
  break :erro_intervalo if f(xi) * f(xf) > 0
  tabela 'k', 'xi', 'xf', 'xm', 'f(xm)'
  iterar do |k|
    xm = (xi + xf) / 2
    fxm = f(xm)
    tabela k, xi, xf, xm, fxm
    break xm if (xf - xi).abs < tol and fxm.abs < tol
    xi, xf = f(xm) * f(xi) < 0 ? [xi, xm] : [xm, xf]
  end
end

File: python/trab3/bissecao.py
from base import *


@algoritmo
def bissecao(self, f, xi, xf):
    self.tabela('k', 'xi', 'xf', 'xm', 'f(xm)')
    if f(xi) * f(xf) > 0:
        raise ErroIntervalo
    for k in xrange(self.nmi):
        xm = (xi + xf) / 2
        fxm = f(xm)
        self.tabela(k, xi, xf, xm, fxm)
        if abs(xf - xi) < self.tol and abs(fxm):
            return xm
        (xi, xf) = (xi, xm) if f(xm) * f(xi) < 0 else (xm, xf)
    raise ErroNMI

